tiledb/sm/query/ast/query_ast.cc

```diff
@@ -0,0 +1,143 @@
+    const tdb_unique_ptr<ASTNode>& rhs,
+    QueryConditionCombinationOp combination_op) {
+  // AST Construction.
+  std::vector<tdb_unique_ptr<ASTNode>> ast_nodes;
+  if (lhs->get_tag() == ASTNodeTag::VAL) {
+    if (rhs->get_tag() == ASTNodeTag::VAL) {
+      ast_nodes.push_back(lhs->clone());
+      ast_nodes.push_back(rhs->clone());
+    } else {
+      // lhs is a simple tree, rhs is a compound tree.
+      auto rhs_tree_expr = dynamic_cast<ASTNodeExpr*>(rhs.get());
+      ast_nodes.push_back(lhs->clone());
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    }
+  } else if (rhs->get_tag() == ASTNodeTag::VAL) {
+    // lhs is a compound tree, rhs is a simple tree.
+    auto lhs_tree_expr = dynamic_cast<ASTNodeExpr*>(lhs.get());
+    for (const auto& elem : lhs_tree_expr->nodes_) {
+      ast_nodes.push_back(elem->clone());
+    }
+    ast_nodes.push_back(rhs->clone());
+  } else if (
+      lhs->get_tag() == ASTNodeTag::EXPR &&
+      rhs->get_tag() == ASTNodeTag::EXPR) {
+    // Both trees are expression trees.
+    auto lhs_tree_expr = dynamic_cast<ASTNodeExpr*>(lhs.get());
+    auto rhs_tree_expr = dynamic_cast<ASTNodeExpr*>(rhs.get());
+
+    if (combination_op == lhs_tree_expr->combination_op_ &&
+        lhs_tree_expr->combination_op_ == rhs_tree_expr->combination_op_) {
+      // The combination ops of the lhs compound tree, rhs compound tree, and
+      // the combination op argument passed in are identical.
+      for (const auto& elem : lhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    } else if (combination_op == lhs_tree_expr->combination_op_) {
+      // The combination op argument passed in is the same as the combination op
+      // of the lhs compound tree.
+      for (const auto& elem : lhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+      ast_nodes.push_back(rhs->clone());
+    } else if (combination_op == rhs_tree_expr->combination_op_) {
+      // The combination op argument passed in is the same as the combination op
+      // of the rhs compound tree.
+      ast_nodes.push_back(lhs->clone());
+      for (const auto& elem : rhs_tree_expr->nodes_) {
+        ast_nodes.push_back(elem->clone());
+      }
+    } else {
+      ast_nodes.push_back(lhs->clone());
+      ast_nodes.push_back(rhs->clone());
+    }
+  } else {
+    throw std::logic_error(
+        "ast_combine: argument node does not have tag VAL or EXPR.");
+  }
+  return tdb_unique_ptr<ASTNode>(
+      tdb_new(ASTNodeExpr, std::move(ast_nodes), combination_op));
+}
+
+void ast_get_field_names(
+    std::unordered_set<std::string>& field_name_set,
+    const tdb_unique_ptr<ASTNode>& node) {
+  if (node == nullptr)
+    return;
+  if (node->get_tag() == ASTNodeTag::VAL) {
+    auto val_ptr = dynamic_cast<ASTNodeVal*>(node.get());
```

Comment:
I think while it is possible to do most of the functions without dynamic cast, it's impossible to do `ast_combine` without dynamic casting, since we don't know the tag of the original node and we need that information to combine the nodes correctly.

Reply:
Sure, I can definitely help you out on that, but I need to understand better what you would like this function to do. Can you give me a complex example with input/output query conditions of this function? 

A few hints: 
- What confuses me in the current implementation is that in the case of type `ASTNodeTag::EXPR` you have a `for` loop that calls `clone()` on each `node` in `nodes_`. Isn't this what just calling `clone()` on an `ASTNodeExpr` object does?
- Also why would you need a `dynamic_cast` since you are calling `clone()` that exists for both subclasses and on top of that you can check from the tag the type of class of each object?

We can sync up offline if you want to discuss more.

Reply:
Deleted all instances of dynamic cast!